lib/cluster.rb

```diff
@@ -176,4 +176,15 @@ def large?
   def empty?
     ht_items.empty? && ocn_resolutions.empty? && holdings.empty? && commitments.empty?
   end
+
+  def update_ocns
+    self.ocns = [component_ocns].flatten.uniq
+    save
+  end
+
+  def component_ocns
+    @component_ocns ||= ocn_resolutions.pluck(:ocns) + ht_items.pluck(:ocns) +
```

Comment:
I think I originally intended for this to be used by `OCNGraph.new.compile_graph_from_cluster` and then rewrote it there anyway. As written, `ocns_changed?` is bugged because it returns true if removed_ocns is a subset of a component's ocns tuple, e.g. ([3] - [1, [2, 3]]).any? is true. I wrote a failing test and then corrected it in ocns_changed?. 

"I suspect that if ocns_changed? is true (i.e. one of the OCNs in the removed clusterable was the last witness to that OCN in the cluster) that we don't need to recluster"
The implementation of ocns_changed? here implies just the opposite since false means there exists an exact duplicate tuple in the cluster. Corrected, ocns_changed? is true is insufficient to determine the need for reclustering because removed tuples can contain more than 2 OCNs. Given a removed tuple of [b, c, e], the first two OCNs can be the glue, and the third the last witness requiring an update to the cluster.ocns.

Reply:
Good point, I wasn't thinking about cases of > 2 OCNs.